core/src/main/java/io/hyperfoil/core/impl/OpenModelPhase.java

```diff
@@ -0,0 +1,101 @@
+package io.hyperfoil.core.impl;
+
+import io.hyperfoil.api.config.Model;
+import io.hyperfoil.api.config.Phase;
+import io.hyperfoil.api.session.Session;
+import io.hyperfoil.core.impl.user.UserGenerator;
+import io.netty.util.concurrent.EventExecutorGroup;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This is a base class for Open Model phases that need to compensate users based on the available ones in the session pool.
+ * <br>
+ * The notion of being "throttled" users requires some explanation:
+ * <ul>
+ *     <li>a user requires a {@link Session} to run, hence the two terms are used to denote the same concept</li>
+ *     <li>starting a {@link Session} doesn't mean immediate execution, but scheduling a deferred start in the {@link Session#executor()}</li>
+ *     <li>given that {@link Session}s are pooled, being throttled means that no available instances are found by {@link #startNewUsers(long)}</li>
+ *     <li>when a {@link Session} finishes, {@link #notifyFinished(Session)} can immediately restart it if there are
+ *     throttled users, preventing it to be pooled</li>
+ * </ul>
+ * <p>
+ * The last point is crucial because it means that a too small amount of pooled sessions would be "compensated" only
+ * when a session finished, instead of when {@link #sessionPool} has available sessions available.
+ */
+final class OpenModelPhase extends PhaseInstanceImpl {
+
+    private final int maxSessions;
+    private final AtomicLong throttledUsers = new AtomicLong(0);
+    private final UserGenerator userGenerator;
+
+    OpenModelPhase(UserGenerator rateGenerator, Phase def, String runId, int agentId) {
+        super(def, runId, agentId);
+        this.userGenerator = rateGenerator;
+        this.maxSessions = Math.max(1, def.benchmark().slice(((Model.OpenModel) def.model).maxSessions, agentId));
+    }
+
+    @Override
+    public void reserveSessions() {
+        log.debug("Phase {} reserving {} sessions", def.name, maxSessions);
+        sessionPool.reserve(maxSessions);
+    }
+
+    @Override
+    public void proceed(final EventExecutorGroup executorGroup) {
+        if (status.isFinished()) {
+            return;
+        }
+        // TODO: epoch from ms is still subject to NTP drifts: maybe we should do something here to account for it!
+        long fireTimeMs = System.currentTimeMillis();
+        long elapsedTimeMs = fireTimeMs - absoluteStartTime;
+        var nextUser = userGenerator.scheduleNextUser(elapsedTimeMs);
```

Comment:
I am not sure I like the name `nextUser` as it might be confusing given that there are possibly more than one user scheduled. Maybe something like `nextScheduledSlot`.. I am not that good at giving names TBH :smile:

Reply:
I've added a new commit which rename the most of the rate generator classes but not the gigantic javadoc I made (I'll wait till the end for such).
Now the behaviour should be 1:1 with the previous one (or better)